File: 算法/BinarySearch/BinarySearch.py
# 该模板的好处
# 普适性 & 通用性
# 循环条件结束清晰 left + 1 < right
# left / right = mid 不用向加1不加1
# 不会死循环

class Solution:
	def binary_search(self, nums: List[int], target: int) -> int:
		if len(nums) == 0:
			return -1
		# 1. initialize
		left = 0
		right = len(nums) - 1

		# 2. 循环结束条件 left + 1 == right
		while left + 1 < right:
			# 不要用(left + right)//2?
			mid = left + (right - left) // 2
			if nums[mid] == target:
				return mid
			elif nums[mid] < target:
				left = mid # 直接等于mid
			else: 
				right = mid # 直接等于mid

		# 3. 出循环left + 1 == right 再判断这两个特殊的case
		if nums[left] == target:
			return left
		elif nums[right] == target:
			return right
		else:
			return -1 # 没有找到

File: BinarySearch/代码随想录示例/区间左闭右开/BinarySearch.java
package 代码随想录示例.区间左闭右开;

public class BinarySearch {

}

class Solution {
    public int search(int[] nums, int target) {
        if (nums.length == 0) {
            return -1;
        }

        int left = 0;
        int right = nums.length - 1;
        // 因为现在要验证左闭右开的情况
        // 所以left和right肯定不能相等
        while (left < right) {
            int mid = (left + right) / 2;
            if (nums[mid] > target) {
                // 这里就得包括上mid的情况
                // 因为right本来就没有包括在这个"左闭右开"的情况下
                // 所以需要包括上mid
                right = mid;
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;
    }
}
